service/worker/scanner/executions/common/types.go

```diff
@@ -0,0 +1,193 @@
+	CheckResultTypeFailed CheckResultType = "failed"
+	// CheckResultTypeCorrupted indicates check successfully ran and detected a corruption
+	CheckResultTypeCorrupted = "corrupted"
+	// CheckResultTypeHealthy indicates check successfully ran and detected no corruption
+	CheckResultTypeHealthy = "healthy"
+
+	// FixResultTypeSkipped indicates that fix skipped execution
+	FixResultTypeSkipped FixResultType = "skipped"
+	// FixResultTypeFixed indicates that fix successfully fixed an execution
+	FixResultTypeFixed = "fixed"
+	// FixResultTypeFailed indicates that fix attempted to fix an execution but failed to do so
+	FixResultTypeFailed = "failed"
+
+	// HistoryExistsInvariantType asserts that history must exist if concrete execution exists
+	HistoryExistsInvariantType InvariantType = "history_exists"
+	// ValidFirstEventInvariantType asserts that the first event in a history must be of a specific form
+	ValidFirstEventInvariantType = "valid_first_event"
+	// PairedWithCurrentInvariantType asserts that an open concrete execution must have a valid current execution
+	PairedWithCurrentInvariantType = "paired_with_current"
+)
+
+// The following are types related to Invariant.
+type (
+	// Execution is an execution which should be checked or fixed.
+	Execution struct {
+		ShardID     int
+		DomainID    string
+		WorkflowID  string
+		RunID       string
+		BranchToken []byte
+		State       int
+	}
+
+	// InvariantResourceBag is a union of resources used to pass results from one Invariant to another Invariant.
+	InvariantResourceBag struct {
+		History *persistence.InternalReadHistoryBranchResponse
+	}
+
+	// CheckResult is the result of running Check.
+	CheckResult struct {
+		CheckResultType CheckResultType
+		Note            string
+		Details         string
+	}
+
+	// FixResult is the result of running Fix.
+	FixResult struct {
+		FixResultType FixResultType
+		Note          string
+		Details       string
+	}
+
+	// ExecutionIteratorResult is a result from ExecutionIterator.Next.
+	ExecutionIteratorResult struct {
+		Execution Execution
+		Error     error
+	}
+)
+
+// The following are serializable types that represent the reports returns by Scan and Fix.
+type (
+	// ShardScanReport is the report of running Scan on a single shard.
+	ShardScanReport struct {
+		ShardID int
+		Handled ShardScanHandled
+		Result  ShardScanResult
+	}
+
+	// ShardScanHandled indicates the executions which were handled by shard Scan.
+	ShardScanHandled struct {
+		ExecutionsCount  int64
+		CorruptedCount   int64
+		CheckFailedCount int64
+		CorruptionByType map[string]int64
+		CorruptOpenCount int64
+	}
+
+	// ShardScanResult indicates the result of running scan on a shard.
+	// Exactly one of ControlFlowFailure or ShardScanKeys will be non-nil
+	ShardScanResult struct {
+		ShardScanKeys      *ShardScanKeys
+		ControlFlowFailure *ControlFlowFailure
+	}
+
+	// ShardScanKeys are the keys to the blobs that were uploaded during scan.
+	ShardScanKeys struct {
```

Comment:
If the activity is restarted, how will we figure out where we left off? I think one way is to look at both of these files, compare the "latest" execution for each type, pick the greatest one, and resume from there. However, that wouldn't work very well since we expect the vast majority of executions to be in a healthy state. It seems like we need a way to remember the last seen healthy execution, too.

Reply:
It won't know where it left off these activities will not heartbeat they will just start the shard over.